src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -864,7 +979,38 @@ public override BoundNode VisitCall(BoundCall node)
                 receiver = node.ReceiverOpt;
                 RefKind refKind = ReceiverSpillRefKind(receiver);
 
+                Debug.Assert(refKind == RefKind.None || !receiver.Type.IsReferenceType);
+
                 receiver = Spill(receiverBuilder, VisitExpression(ref receiverBuilder, receiver), refKind: refKind);
+
+                if (refKind != RefKind.None &&
+                    CodeGenerator.IsPossibleReferenceTypeReceiverOfConstrainedCall(receiver) &&
+                    !CodeGenerator.ReceiverIsKnownToReferToTempIfReferenceType(receiver) &&
+                    !CodeGenerator.IsSafeToDereferenceReceiverRefAfterEvaluatingArguments(node.Arguments))
+                {
+                    var receiverType = receiver.Type;
+                    Debug.Assert(!receiverType.IsReferenceType);
+
+                    // A case where T is actually a class must be handled specially.
+                    // Taking a reference to a class instance is fragile because the value behind the 
+                    // reference might change while arguments are evaluated. However, the call should be
+                    // performed on the instance that is behind reference at the time we push the
+                    // reference to the stack. So, for a class we need to emit a reference to a temporary
+                    // location, rather than to the original location
+
+                    // If condition `(object)default(T) != null` is true at execution time,
+                    // the T is a value type. And it is a reference type otherwise.
+                    var isValueTypeCheck = _F.ObjectNotEqual(
+                                                _F.Convert(_F.SpecialType(SpecialType.System_Object), _F.Default(receiverType)),
+                                                _F.Null(_F.SpecialType(SpecialType.System_Object)));
+
+                    var cache = _F.Local(_F.SynthesizedLocal(receiverType));
+                    receiverBuilder.AddLocal(cache.LocalSymbol);
+                    receiverBuilder.AddStatement(_F.If(_F.Not(isValueTypeCheck), _F.Assignment(cache, receiver)));
```

Comment:
> Should we be asserting that this assignment is in line with ```IsComplexConditionalInitializationOfReceiverRef```, like we do in ```LocalRewriter_Call```?

It is not in line with that and the assert would fail. An assignment that we create here doesn't have a special shape and we don't need to treat it specially anywhere.